extensions/MongoDbAtlas/MongoDbAtlas/MongoDbAtlasMemory.cs

```diff
@@ -0,0 +1,351 @@
+        // score = (1 + cosine/dot_product(v1,v2)) / 2
+        // Thus it does not output the real cosine similarity, this is annoying so we
+        // need to recompute cosine similarity using the embeddings.
+#if DEBUG
+        //List<(double, double, double)> similarities = new();
+#endif
+        foreach (var document in documents)
+        {
+            var memoryRecord = FromMongodbMemoryRecord(document, withEmbeddings);
+
+            // we have score that is normalized, so we need to recompute similarity to have a real cosine distance
+            var cosineSimilarity = CosineSim(embeddings, document.Embedding);
+            if (cosineSimilarity < minRelevance)
+            {
+                //we have reached the limit for minimum relevance so we can stop iterating
+                break;
+            }
+#if DEBUG
+            // Tried to calculate cosine similarity thanks to helper function of embeddings class but the precision
+            // is different from what is used by the tests, so test fails. This forces me to use actually the
+            // CosineSim formula. This commented code can be use in debugging to verify the differences between
+            // various values
+            // var score = document["score"].AsDouble;
+            // var cosineSimilarity2 = embeddings.CosineSimilarity(memoryRecord.Vector);
+            // similarities.Add((score, cosineSimilarity, cosineSimilarity2));
+            // var diff = Math.Abs(cosineSimilarity - cosineSimilarity2);
+#endif
+            yield return (memoryRecord, cosineSimilarity);
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<string> UpsertAsync(string index, MemoryRecord record, CancellationToken cancellationToken = default)
+    {
+        var normalizedIndexName = NormalizeIndexName(index);
+        var collection = this.GetCollectionFromIndexName(index);
+        MongoDbAtlasMemoryRecord mongoRecord = new()
+        {
+            Id = record.Id,
+            Index = normalizedIndexName,
+            Embedding = record.Vector.Data.ToArray(),
+            Tags = record.Tags.Select(x => new MongoDbAtlasMemoryRecord.Tag(x.Key, x.Value.ToArray())).ToList(),
+            Payloads = record.Payload.Select(x => new MongoDbAtlasMemoryRecord.Payload(x.Key, x.Value)).ToList()
+        };
+
+        await collection.InsertOneAsync(mongoRecord, cancellationToken: cancellationToken).ConfigureAwait(false);
+        await this.Config.AfterIndexCallbackAsync().ConfigureAwait(false);
+        return record.Id;
+    }
+
+    private FilterDefinition<MongoDbAtlasMemoryRecord>? TranslateFilters(ICollection<MemoryFilter>? filters, string index)
+    {
+        List<FilterDefinition<MongoDbAtlasMemoryRecord>> outerFiltersArray = new();
+        foreach (var filter in filters ?? Array.Empty<MemoryFilter>())
+        {
+            var thisFilter = filter.GetFilters().ToArray();
+            var numOfFilters = thisFilter.Count(x => !string.IsNullOrEmpty(x.Value));
```

Comment:
this var is never used

Reply:
you are right, it is a leftover for some tests I made at the beginning.